File: jarvis-personal/native/ios/DincrKit/Sources/DincrCore/MoneyFormat.swift
import Foundation

/// Presentation-only money formatting from the user's profile preferences
/// (`base_currency`, `number_format`, `currency_placement` set in profile setup).
/// It never converts between currencies. Display rounding (colones without decimals) is visual
/// only: editable text uses `inputText`, which keeps every stored digit.
public struct MoneyFormat: Sendable, Equatable {
    public enum Separators: String, Sendable { case dotComma = "dot_comma", commaDot = "comma_dot" }
    public enum Placement: String, Sendable { case before, after }
    public enum Sign: Sendable { case none, income, expense }

    public let currency: String
    public let separators: Separators
    public let placement: Placement

    public init(currency: String = "CRC", separators: Separators = .dotComma, placement: Placement = .before) {
        self.currency = currency.uppercased(); self.separators = separators; self.placement = placement
    }

    public init(profile: Profile?) {
        self.init(
            currency: profile?.baseCurrency ?? "CRC",
            separators: Separators(rawValue: profile?.numberFormat ?? "") ?? .dotComma,
            placement: Placement(rawValue: profile?.currencyPlacement ?? "") ?? .before
        )
    }

    public var symbol: String {
        switch currency {
        case "CRC": "₡"
        case "USD": "$"
        case "EUR": "€"
        default: currency
        }
    }

    /// Colones are shown without decimals (as in the Capacitor app); other currencies with 2.
    public var fractionDigits: Int { currency == "CRC" ? 0 : 2 }

    public func string(_ amount: Decimal, sign: Sign = .none, currency override: String? = nil) -> String {
        let format = override.map { MoneyFormat(currency: $0, separators: separators, placement: placement) } ?? self
        let magnitude = format.digits(abs(amount))
        // A currency code (legacy bases such as ARS) is a word, so it gets a space: "ARS 1.234,00".
        let gap = format.symbol.count > 1 ? " " : ""
        let body = format.placement == .before ? "\(format.symbol)\(gap)\(magnitude)" : "\(magnitude) \(format.symbol)"
        let prefix: String = switch sign {
        case .income: "+"
        case .expense: "−"
        case .none: amount < 0 ? "−" : ""
        }
        return prefix + body
    }

    /// A complete phrase for VoiceOver, e.g. "menos 18.450 colones". `currency` names the unit
    /// when the amount is not in the base currency, exactly like `string(_:sign:currency:)`.
    public func spoken(_ amount: Decimal, sign: Sign = .none, currency override: String? = nil, language: AppLanguage = .current) -> String {
        if let override, override.uppercased() != currency {
            return MoneyFormat(currency: override, separators: separators, placement: placement).spoken(amount, sign: sign, language: language)
        }
        let negative = sign == .expense || (sign == .none && amount < 0)
        let positive = sign == .income
        // Screen readers read display grouping literally ("257.550" can become "257 point 55" in
        // English), so the spoken form uses ungrouped digits and the language's decimal mark.
        let number = spokenDigits(abs(amount), language: language)
        let unit = unitName(language: language, plural: abs(amount) != 1)
        let lead = negative ? language.pick("menos ", "minus ") : positive ? language.pick("más ", "plus ") : ""
        return "\(lead)\(number) \(unit)"
    }

    /// The stored amount as the user would type it (their separators, no symbol, no rounding),
    /// for prefilling an edit form. `AmountInput.parse` reads it back to the same value.
    public func inputText(_ amount: Decimal) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.minimumFractionDigits = 0
        formatter.maximumFractionDigits = AmountInput.maxFractionDigits
        formatter.roundingMode = .halfUp
        formatter.usesGroupingSeparator = true
        formatter.groupingSize = 3
        formatter.groupingSeparator = separators == .dotComma ? "." : ","
        formatter.decimalSeparator = separators == .dotComma ? "," : "."
        return formatter.string(from: abs(amount) as NSDecimalNumber) ?? "\(abs(amount))"
    }

    func digits(_ value: Decimal) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.minimumFractionDigits = fractionDigits
        formatter.maximumFractionDigits = fractionDigits
        // Half away from zero, like the Capacitor app's Intl.NumberFormat (₡2,5 → ₡3).
        formatter.roundingMode = .halfUp
        formatter.usesGroupingSeparator = true
        formatter.groupingSize = 3
        formatter.groupingSeparator = separators == .dotComma ? "." : ","
        formatter.decimalSeparator = separators == .dotComma ? "," : "."
        return formatter.string(from: value as NSDecimalNumber) ?? "\(value)"
    }

    func spokenDigits(_ value: Decimal, language: AppLanguage) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.roundingMode = .halfUp
        formatter.usesGroupingSeparator = false
        formatter.minimumFractionDigits = 0
        formatter.maximumFractionDigits = fractionDigits
        formatter.decimalSeparator = language == .spanish ? "," : "."
        return formatter.string(from: value as NSDecimalNumber) ?? "\(value)"
    }

    func unitName(language: AppLanguage, plural: Bool) -> String {
        switch (currency, language) {
        case ("CRC", .spanish): plural ? "colones" : "colón"
        case ("CRC", .english): plural ? "colones" : "colón"
        case ("USD", .spanish): plural ? "dólares" : "dólar"
        case ("USD", .english): plural ? "dollars" : "dollar"
        case ("EUR", _): plural ? "euros" : "euro"
        default: currency
        }
    }
}

/// Parses what the user typed in an amount field, following their separator preference.
/// Returns nil for anything that is not an unambiguous positive amount: a wrong value is worse
/// than asking again, so nothing is guessed.
public enum AmountInput {
    /// Money never has more than two decimals. It also makes "1.000" (comma_dot) and "1,000"
    /// (dot_comma) invalid instead of silently meaning one.
    public static let maxFractionDigits = 2
    /// Twelve integer digits (below a trillion) is far above any personal amount and keeps
    /// accidental pastes out of the ledger.
    public static let maxIntegerDigits = 12

    public static func parse(_ text: String, separators: MoneyFormat.Separators) -> Decimal? {
        let trimmed = text.replacingOccurrences(of: " ", with: "")
        guard !trimmed.isEmpty else { return nil }
        let grouping: Character = separators == .dotComma ? "." : ","
        let decimal: Character = separators == .dotComma ? "," : "."
        guard trimmed.allSatisfy({ $0.isASCII && ($0.isNumber || $0 == grouping || $0 == decimal) }) else { return nil }
        let halves = trimmed.split(separator: decimal, omittingEmptySubsequences: false)
        guard halves.count <= 2 else { return nil }
        let integer = halves[0]
        let fraction = halves.count == 2 ? halves[1] : ""
        guard !integer.isEmpty, !fraction.contains(grouping), fraction.count <= maxFractionDigits else { return nil }
        // Grouping must be real thousands groups ("18.450"), never a mistyped decimal
        // ("1,5" in comma_dot would otherwise silently become 15).
        let groups = integer.split(separator: grouping, omittingEmptySubsequences: false)
        if groups.count > 1 {
            guard let first = groups.first, (1...3).contains(first.count),
                  groups.dropFirst().allSatisfy({ $0.count == 3 }) else { return nil }
        }
        let integerDigits = groups.joined().drop(while: { $0 == "0" })
        guard integerDigits.count <= maxIntegerDigits else { return nil }
        let digits = groups.joined() + (fraction.isEmpty ? "" : "." + fraction)
        guard let value = Decimal(string: digits, locale: Locale(identifier: "en_US_POSIX")), value > 0 else { return nil }
        return value
    }
}

File: jarvis-personal/native/android/core/data/src/main/kotlin/com/dincr/data/MoneyFormat.kt
package com.dincr.data

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Presentation-only money formatting from the user's profile (`base_currency`, `number_format`,
 * `currency_placement`). Never converts currencies. Same outputs as iOS `MoneyFormat`.
 */
data class MoneyFormat(
    val currency: String = "CRC",
    val separators: Separators = Separators.DOT_COMMA,
    val placement: Placement = Placement.BEFORE,
) {
    enum class Separators(val wire: String) { DOT_COMMA("dot_comma"), COMMA_DOT("comma_dot") }
    enum class Placement(val wire: String) { BEFORE("before"), AFTER("after") }
    enum class Sign { NONE, INCOME, EXPENSE }

    val symbol: String get() = when (currency.uppercase()) { "CRC" -> "₡"; "USD" -> "$"; "EUR" -> "€"; else -> currency.uppercase() }
    val fractionDigits: Int get() = if (currency.uppercase() == "CRC") 0 else 2

    fun format(amount: BigDecimal, sign: Sign = Sign.NONE, currencyOverride: String? = null): String {
        val f = currencyOverride?.let { copy(currency = it) } ?: this
        val magnitude = f.digits(amount.abs())
        val body = if (f.placement == Placement.BEFORE) "${f.symbol}$magnitude" else "$magnitude ${f.symbol}"
        val prefix = when (sign) {
            Sign.INCOME -> "+"
            Sign.EXPENSE -> "−"
            Sign.NONE -> if (amount.signum() < 0) "−" else ""
        }
        return prefix + body
    }

    /** A complete phrase for TalkBack, e.g. "menos 18.450 colones". */
    fun spoken(amount: BigDecimal, sign: Sign = Sign.NONE, language: AppLanguage = AppLanguage.current()): String {
        val negative = sign == Sign.EXPENSE || (sign == Sign.NONE && amount.signum() < 0)
        val lead = when {
            negative -> language.pick("menos ", "minus ")
            sign == Sign.INCOME -> language.pick("más ", "plus ")
            else -> ""
        }
        val plural = amount.abs().compareTo(BigDecimal.ONE) != 0
        val unit = when (currency.uppercase() to language) {
            "CRC" to AppLanguage.SPANISH, "CRC" to AppLanguage.ENGLISH -> if (plural) "colones" else "colón"
            "USD" to AppLanguage.SPANISH -> if (plural) "dólares" else "dólar"
            "USD" to AppLanguage.ENGLISH -> if (plural) "dollars" else "dollar"
            else -> currency.uppercase()
        }
        // Screen readers read display grouping literally ("257.550" can become "257 point 55" in
        // English), so the spoken form uses ungrouped digits and the language's decimal mark.
        val plain = amount.abs().setScale(fractionDigits, RoundingMode.HALF_EVEN).stripTrailingZeros().toPlainString()
        val spokenNumber = if (language == AppLanguage.SPANISH) plain.replace('.', ',') else plain
        return "$lead$spokenNumber $unit"
    }

    internal fun digits(value: BigDecimal): String {
        val scaled = value.setScale(fractionDigits, RoundingMode.HALF_EVEN).toPlainString()
        val integer = scaled.substringBefore('.')
        val fraction = scaled.substringAfter('.', "")
        val group = if (separators == Separators.DOT_COMMA) '.' else ','
        val decimal = if (separators == Separators.DOT_COMMA) ',' else '.'
        val grouped = integer.reversed().chunked(3).joinToString(group.toString()).reversed()
        return if (fraction.isEmpty()) grouped else "$grouped$decimal$fraction"
    }

    companion object {
        fun from(profile: Profile?) = MoneyFormat(
            currency = profile?.baseCurrency ?: "CRC",
            separators = Separators.entries.firstOrNull { it.wire == profile?.numberFormat } ?: Separators.DOT_COMMA,
            placement = Placement.entries.firstOrNull { it.wire == profile?.currencyPlacement } ?: Placement.BEFORE,
        )
    }
}

/**
 * Parses typed amounts. Grouping must be real thousands groups ("18.450"); a mistyped decimal
 * ("1,5" in comma_dot) is rejected instead of silently becoming 15.
 */
object AmountInput {
    fun parse(text: String, separators: MoneyFormat.Separators): BigDecimal? {
        val trimmed = text.replace(" ", "")
        if (trimmed.isEmpty()) return null
        val group = if (separators == MoneyFormat.Separators.DOT_COMMA) '.' else ','
        val decimal = if (separators == MoneyFormat.Separators.DOT_COMMA) ',' else '.'
        if (!trimmed.all { it in '0'..'9' || it == group || it == decimal }) return null
        val halves = trimmed.split(decimal)
        if (halves.size > 2) return null
        val integer = halves[0]
        val fraction = halves.getOrElse(1) { "" }
        if (integer.isEmpty() || fraction.contains(group)) return null
        val groups = integer.split(group)
        if (groups.size > 1 && (groups.first().length !in 1..3 || groups.drop(1).any { it.length != 3 })) return null
        val value = (groups.joinToString("") + if (fraction.isEmpty()) "" else ".$fraction").toBigDecimalOrNull() ?: return null
        return value.takeIf { it.signum() > 0 }
    }
}
